### DISCUSSION TITLE
Welcome to trs80-diagnosticrom Discussions!

### DISCUSSION THREAD

**Discussion by misterblack1** [misterblack1/trs80-diagnosticrom#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by RadioC1ash** [comment]
I've got a Model 4, some EPROMS, an adapter, a programmer, and no sense of respect for my limitations. What can I do to help with creating and validating a Model 4 diagnostic ROM?

I'm in corporate IT, don't really have a coding background, though. I've been working with and repairing computer hardware since the early 90's. I've got a basement full of projects and a boneyard of electronic components that takes up all my spare time. That pretty well sums it up!

**Reply by RadioC1ash** [reply]
![IMG_9385](https://github.com/misterblack1/trs80-diagnosticrom/assets/131454766/8f77fec3-b799-46ab-ad1c-6c4fe15805bc)
I was able to install the Model 1/3 version of the diagnostic ROM in my Model 4 (NGA) using the adapters built from the linked PCBWay shared community project. It seems to run just as if it were installed in a Model 3, which is probably as expected. Anyhow, that’s the state of things on my bench at the minute.

**Reply by ki3v** [reply]
Hi!  Please give the Model 4P ROM a try in your model 4.  It's untested on the M4, but per the hardware manuals I had access to when I wrote the 4P version, it should work on the 4 and 4D as well.  You'll be the first person to try it out on M4 hardware if you do.

The M4P test should work on the M4... we just called it the M4P version because we only had an M4P to test with (the one in Adrian's video).

The M3 test will work as you saw (thanks for testing it!), but the banking information is wrong (The 4/4D/4P have one or two 64K banks, not three 16K banks like the M3) and it won't test all of the M4 RAM.  The M3 test will only test 48K (from 3000-FFFF), but the 4P version will test up to 128K in machines that have the expanded RAM.

(Actually, looking at the GIF animation on the code page... I can't remember if it tests the bottom 16K that's initially hidden by the M4 ROM... No, skimming the source I see that I didn't.  I would have had to perform the trick of relocating the RAM test code into upper RAM and then banking in the lowest 16K of RAM and it looks like that wasn't finished in the 4/4D/4P version...)

If I recall it will also initialize your video in 80 column mode.  The text will be on the left side of the screen (the first 64 columns) during the actual memory test, but the initial VRAM test will fill the display for a few seconds as an aid to hardware troubleshooting.

**Reply by RadioC1ash** [reply]
![IMG_9386](https://github.com/misterblack1/trs80-diagnosticrom/assets/131454766/3666ac62-a7e2-4297-b5a9-e2ff8e7f7f2c)
Well, it does run the Model 4P version. This is as good a time as any, I suppose, to mention that I don’t know how well my machine is working. That said, everything seems to work reasonably well in day to day use. It’s a result, in any case.

**Reply by RadioC1ash** [reply]
At one point it changed a bit. 4000-FFFF read “absent" briefly. That hasn’t showed up again yet. 
Also, it does fill the screen briefly with characters as described. 
![IMG_9393](https://github.com/misterblack1/trs80-diagnosticrom/assets/131454766/9419b5ad-927e-4668-8cd0-642bcfdd38a9)


![IMG_9387](https://github.com/misterblack1/trs80-diagnosticrom/assets/131454766/ed9c0582-5bc4-431e-96b2-52e8c90f9848)

**Comment by rarabian** [comment]
Hello,

I'm trying to use this rom in my Model 4 GA version and getting strange results.  A screen full of 40's.  I've tested all of the ram and even swapped the 4th chip based on the pictures with the same result.  It takes a few seconds to switch from random stuff to 40's but never goes beyond that. The machine has 128k of ram. I've pulled out the upper bank and nothing changed. Here's a short video that shows the behavior:

https://github.com/misterblack1/trs80-diagnosticrom/assets/8408653/0a244d6d-6a61-4707-b8a1-16afe6453ed0

Any advice you can give would be appreciated.

Thanks,

--Rick
